core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -361,29 +362,68 @@
         (vr/<-root vsr))))
 
   (test-metadata [_mpg]
-    (.test page-idx-pred page-idx)))
+    (.test page-idx-pred page-idx))
+
+  (temporal-bounds [_mpg] (.temporalBounds table-metadata (int page-idx))))
+
+(def ^:private non-constraint-bounds (TemporalBounds.))
 
 (defrecord LiveMergePlanPage [^RelationReader live-rel trie ^LiveHashTrie$Leaf leaf]
   MergePlanPage
   (load-page [_mpg _buffer-pool _vsr-cache]
     (.select live-rel (.mergeSort leaf trie)))
 
-  (test-metadata [_mpg] true))
-
-(defn ->merge-task [mp-pages]
-  (let [leaves (ArrayList.)]
-    (loop [[mp-page & more-mp-pages] mp-pages
-           node-taken? false]
-      (if mp-page
-        (let [take-node? (test-metadata mp-page)]
-
-          (when (or take-node? node-taken?)
-            (.add leaves mp-page))
-
-          (recur more-mp-pages (or node-taken? take-node?)))
-
-        (when node-taken?
-          (vec leaves))))))
+  (test-metadata [_mpg] true)
+
+  (temporal-bounds [_msg] non-constraint-bounds))
+
+(defn max-valid-to [^TemporalBounds tb]
+  (.getUpper (.getValidTo tb)))
+
+(defn min-valid-from [^TemporalBounds tb]
+  (.getLower (.getValidFrom tb)))
+
+(defn ->merge-task
+  ([mp-pages] (->merge-task mp-pages (TemporalBounds.)))
+  ([mp-pages ^TemporalBounds query-bounds]
+   (let [leaves (ArrayList.)]
+     (loop [[mp-page & more-mp-pages] mp-pages
+            node-taken? false
+            largest-valid-to Long/MIN_VALUE
+            non-taken-pages []]
+       (if mp-page
+         (let [^TemporalBounds page-bounds (temporal-bounds mp-page)
+               take-node? (and (.intersects page-bounds query-bounds)
+                               (test-metadata mp-page))]
+
+           (if take-node?
+             (do
+               (.add leaves mp-page)
+               (recur more-mp-pages
+                      true
+                      (cond-> largest-valid-to
+                        (not (instance? LiveMergePlanPage mp-page))
+                        (max (max-valid-to page-bounds)))
+                      non-taken-pages))
+
+             (recur more-mp-pages
+                    node-taken?
+                    largest-valid-to
+                    (cond-> non-taken-pages
+                      (and node-taken?
+                           (.intersectsSystemTime page-bounds query-bounds)
+                           (.canBoundValidTime query-bounds page-bounds))
+                      (conj mp-page)))))
+
+         (when node-taken?
+           (loop [[page & more-pages] non-taken-pages]
```

Comment:
These loop could be made conditional on `valid-to` projections under the assumption we allow for version coalescing.